### PR TITLE
More similar mean batch duration across nodes with DynamicBucketingSampler in multi-GPU training

### PR DESCRIPTION
**Pull request by lifeiteng**
Update #863 

Fix #857

### CODE REVIEW

**Comment by pzelasko** [comment]
Does it actually help you achieve greater GPU utilization? Would be great if you can share any benchmarks. I wasn't able to verify the improvements in the original PR.

**Comment by lifeiteng** [comment]
> Does it actually help you achieve greater GPU utilization? Would be great if you can share any benchmarks. I wasn't able to verify the improvements in the original PR.

It will be around 2x faster on 8xGPU.

**Comment by pzelasko** [comment]
> It will be around 2x faster on 8xGPU.

Wow!! That's fantastic! Let me know if you need any help along the way.

File: lhotse/dataset/sampling/dynamic_bucketing.py [hunk 1]
```diff
@@ -427,38 +448,51 @@ def is_ready(bucket: Deque[Cut]):
...
-                    for _ in range(batch_size):
-                        sampling_bucket.popleft()
+                # Shuffling, sort indexes of yielded elements largest -> smallest and remove them
+                indexes_used.sort(reverse=True)
+                for idx in indexes_used:
+                    del sampling_bucket[idx]
+
                 # Fetch new cuts and add them to appropriate buckets.
-                self._collect_cuts_in_buckets(batch_size)
+                self._collect_cuts_in_buckets(total_batch_size * self.world_size)
```
Comment (lifeiteng) [inline lhotse/dataset/sampling/dynamic_bucketing.py:495]: we need collect more cuts here, It is often encountered that there is insufficient data in buckets.  @pzelasko Do you have any suggestion?

Comment (lifeiteng) [reply lhotse/dataset/sampling/dynamic_bucketing.py:495]: sample batches from the some bucket(word_size times), then get batches[rank].

**Pull request closed without merging**